lib/rate_map.c

```diff
@@ -153,25 +157,24 @@ rate_map_position_to_mass(rate_map_t *self, double pos)
 {
     const double *position = self->position;
     const double *rate = self->rate;
-    double offset;
+    double base, offset;
     size_t index;
 
     assert(pos >= 0.0);
-    if (pos <= position[0]) {
+    if (pos <= 0.0) {
         return 0;
     }
     assert(pos <= position[self->size]);
-    if (pos >= position[self->size]) {
-        return self->cumulative_mass[self->size];
-    }
-    /* search for upper bounds strictly before the max position
-       any position greather than or equal to max position returns self->size */
-    index = msp_binary_interval_search(pos, position, self->size);
+    index = idx_1st_strict_upper_bound(position, self->size + 1, pos);
     assert(index > 0);
     index--;
+    base = self->cumulative_mass[index];
     offset = pos - position[index];
-
-    return self->cumulative_mass[index] + offset * rate[index];
+    assert(offset >= 0);
+    if (offset <= 0) {
```

Comment:
I don't see how offset could be < 0 here - aren't we checking that all rates and masses are >= 0 on the way in?

Reply:
I'm getting a little ahead of myself with #1203 in the pipeline but not yet merged. I'll take it out to make merging #1203 easier. I also don't see anyway offset >=0 isn't true in the current impl but that's with some assumptions about the other code and data structures (present and future). Once #1203 is merged there's another discussion which is when is being 'assert happy' too enthusiastically ... 'assertive' ha ha :)